spatial: st_buffer query causes RandomSyntax tests to time out

A spatial query can cause the RandomSyntax tests to time out. See https://teamcity.cockroachdb.com/viewLog.html?buildId=2733079&buildTypeId=Cockroach_Nightlies_RandomSyntaxTests&tab=buildLog&branch_Cockroach_Nightlies=release-20.2&_focus=15567

```
timeout: "SELECT st_buffer('01060000000100000001030000000100000007000000595466C414D8E9C7B8A9196ACC65E4C74A0576084BACDAC7E9A04CB1E011EFC77E701EEF7191E34748E7F0473F3AE8C782CF2146F24FE54706E38BB5CAA5D4C73A321CC4A1A4E047F86753B84FA4E647607EC493E1B7B7477A25F1A25773D1C7595466C414D8E9C7B8A9196ACC65E4C7':::GEOMETRY, 1.7976931348623157e+308:::FLOAT8, (-2791747071544515918):::INT8) ". currently executing: map[SELECT st_buffer('0102000020E610000005000000F0C93AD13A1745C050C8FFCFF66846C0665D16CB897150405F9B43A3329647C08EBF2FB46B8A504000521838EABBD1BF7C230605DBEA56408CD6045457314440140D5C0EA61863C0EAB113C2C8055640':::GEOGRAPHY, 1.0337640987193146:::FLOAT8, (-2468804744006758512):::INT8) :1 SELECT st_buffer('0103000020E610000001000000090000007938722990DE65C0DC78DE3547B036C0D176FAC5BB2264C06C38DC4641023CC08EEE627621CF5AC025F2C3D02FBC52C074732473E76140402007E2B953BD4140405F9D5D1D8300C0C0B80B991E5A504048034464064546C0EC9C8EC316ED4E40F0A612AF5D9660C0482A5FEE4FF14F40D6ACF8E491DA61C068AEEA41EA5942407938722990DE65C0DC78DE3547B036C0':::GEOGRAPHY, (-0.8485197301516284):::FLOAT8, (-1303200549869436442):::INT8) :1 SELECT st_buffer('0103000020E6100000010000000D000000462464CFA55E4BC0C82B40EC65DF22C0BA7B98C5199A65C0EC63B6B1365654C07B67CB10DDEC58C087566C70DBD34CC0386E28AD014352C0E2B74E9666E544C0C9CC1050F7005BC01C14C8C48CBC51C0C011A4785DAA6340803BCB15E94EEFBF923F204EC894604080D4A90ADF6A32408489504A6FE05A40D675723627934240747CCEAF26826240A43AB5AB2A115140000BC075293EEC3F060D1894AFC250408019642F22F049C01C95DA8099665140DD4ECFA8114C5AC0C852E40C555F3040462464CFA55E4BC0C82B40EC65DF22C0':::GEOGRAPHY, 1.1198911720336593:::FLOAT8, (-5609002936553243121):::INT8) :1 SELECT st_buffer('0103000020E6100000020000000500000000000000000000000000000000000000000000000000F03F0000000000000000000000000000F03F000000000000F03F0000000000000000000000000000F03F00000000000000000000000000000000050000009A9999999999C93F9A9999999999C93F9A9999999999C93F9A9999999999D93F9A9999999999D93F9A9999999999D93F9A9999999999D93F9A9999999999C93F9A9999999999C93F9A9999999999C93F':::GEOGRAPHY, 0.9679180840912789:::FLOAT8, (-8468854772686012558):::INT8) :1 SELECT st_buffer('0104000000050000000101000000F8889017E3A5EB470E7B291E2301E6C701010000007C89271223B7DD476038A2624FB8ECC701010000006270CD63EE35ECC7089F3D30F0D8CA4701010000004A983E098F81E24791490A7E0101EDC7010100000072E3A5655AFAE2C7107EA0379F73EE47':::GEOMETRY, 0.4269145048214039:::FLOAT8, (-8057790196575783732):::INT8) :1 SELECT st_buffer('01050000000200000001020000000300000000F65C62D73BE547B5CF51E89B98ECC7861C1301454DEF471CEA9A12A910E3C7E6438F60F2F2E4C784D675784F09D147010200000005000000009863FB3D6FA8C7F9551D22DA01EAC7200AB62A41F3A747EA372FE940EDDCC76EF59933DD21E0479ADAD6CDAB6AD5C71067F733604FE84754F6F20CCFA6D54750FCB69E2ABBCD4704F720FAB64ED247':::GEOMETRY, 5e-324:::FLOAT8, (-3243394533633060744):::INT8) :1 SELECT st_buffer('01060000000100000001030000000100000007000000595466C414D8E9C7B8A9196ACC65E4C74A0576084BACDAC7E9A04CB1E011EFC77E701EEF7191E34748E7F0473F3AE8C782CF2146F24FE54706E38BB5CAA5D4C73A321CC4A1A4E047F86753B84FA4E647607EC493E1B7B7477A25F1A25773D1C7595466C414D8E9C7B8A9196ACC65E4C7':::GEOMETRY, 1.7976931348623157e+308:::FLOAT8, (-2791747071544515918):::INT8) :1 SELECT st_buffer('0107000020E6100000060000000102000020E610000009000000E00DC4F940203B40642B7525006955C02CFFFB34B1D346400B5A676EBE9946C02C4FE549E05B52400AAABA9851E74DC0085199FCC69A544028362F891C3B30C0583A04B528E05B407CADCEA344F532C004E0F1528C246340CAE15C4429BB41402CBE21F7AF286140A0CA2B35065F3D4084BB4B09D5B04040289EE5C5DB994D40015AA8C3EDD45BC050640B36138A39400103000020E610000001000000080000004445F470CD1160C06CE64ADEE4CE37C0584D6B297EFA58C0E8FCA306493B41C001C86E4520EA57C0F06FF17C31C653C0ECB563CD253C5F40E6B42841190750C0000E84A0C671D6BFC6838EFD424540405620FDA5372A4BC0E09FCD3F94E128405445966EE1A251C01CE0A973BB9540404445F470CD1160C06CE64ADEE4CE37C00106000020E61000000500000001030000000100000005000000105D847EBAA260C0AA990D0D49DB45C02621B2509AAC62407E7D6C14719F50C0F0A67B7B56A96540E0F61A74BFA31CC0371B943EFD7956C0D0DE37A3A1915140105D847EBAA260C0AA990D0D49DB45C00103000000010000000D000000A46B9647EF6F5EC0D04673B3A1841AC0F3192C958C4250C0E021CF779FB93CC019CC1F2554BB57C01109B3F296EC51C0D08D60F269243A4063CBCD81C85E4AC0A61632C99D9F624000540A9DCF5EFB3F167B22BE3BBF60403013FAA8E01A2440EC7E6D809D6863405CBD4B9AF27730402293C480A61C5240A45F39E4BBB13340F05CA9648BE027C064856164C84C474074DC02184BF94FC00AFB96B68C8B5440CC9A9F5ED31C49C038418D3E467747400DAC08BF1B4A54C0684BE6B400A42440A46B9647EF6F5EC0D04673B3A1841AC001030000000100000005000000EFF5C202F93E63C0156FC9B2A42453C070A3411549E560C004D0B897E43556C0F87520A89A295B40340D7099698246C0B4D0A9D348E049408CA46297829F4040EFF5C202F93E63C0156FC9B2A42453C00103000000010000000D000000D3585676B63166C0201EA1263369004098671D18EF1F55C0A0EF2A9696FA23C0FE8530ABE18D52404E3C580B971355C0E89A671182285A40AD152989593345C058F3E75FC9A8584000D643084664D2BFFCB2C796D3896340545F02418E823D40946F88D878686640FC39B4F89ED949405656D224804862400CAEA3E4BE774640AC97476621DE5B40A0081CAF8365554034992E9AB3D14840EAAE6F81AF17554060B3882BB0964740D068C78E89BC53404C1189157AD142405AFAAAB00E815540D3585676B63166C0201EA1263369004001030000000100000005000000C4E205B03E2C40C09A1E83515E163CC038231E68E6BB4040D83C8444ECDF2C407C7EC6452D7E634066DF7F2E34DC5240F85FF2739DC23CC01C4940BD9AD94B40C4E205B03E2C40C09A1E83515E163CC00105000020E610000007000000010200000004000000E282684096BD5DC000ECAE496B7FC03FCCF534EE8E7B64C0180AF38B7CC621C024E78F39F1444840486D3E3B37FD34C0608158425E0E3AC0887EE6E7D8A32640010200000003000000CAFB835F2E4D604045AA28018B2A4BC0AC5DB294AE3C56406875ABAA27A94E4021037235F3DD65C09852091A848550400102000000050000003EB0D77DBA704CC0F34B6D94808950C05E87538923725140249330AD66FF36C0A031F1608EBB3940703F2C2DD8ED3140881BBB10A42B4E4030D0CE5AD1F9364002C84FD4372C64C07E0FB8048D375240010200000003000000C0891E38AE3B63C08C38B9C05FAE53C0AF542B2FE60F51C0FFD4D54E1CB74FC0A0ACFCEDAA853240403082830CF40240010200000005000000C009A35732D219C033628A2A1DBB49C0E054E8646D1E1040271D8E28698151C030426E0B118433404730F7A9D15D55C058EB57DCED154440436D2082DA0F42C032FBEBB1DAD86040B8ECE2B43BA826C0010200000008000000991801834AC965C02214CDE3312C54C03E0CA49F6B3756C074FA5DF2D59F53C0AA56BBE33FAD4EC07C53B60C3F8F4FC090C8E206F9D031C09BCE9F0B6B9A55C0DC6FE40195515540DF2C5D1EB0E852C020B1C1DD56B2384085EDAED362CB48C05AE9AF1BB9135240F87314E0B4474A4080102FDE4B285AC040F2308E2D4C2B40010200000003000000C0007074252520C0082A42FC626120C0506995959F3D40C060E1B3DA41CA2FC06423C35BCEFB54407E304859C4F24EC00104000020E61000000400000001010000005697420D6BFF60C02C83AAC2292854C001010000002403736A0ECE40C00AD9073B54763BC00101000000623D07F207C850402A678E40E0D453C00101000000DCBB001BEC4046C084DA13FA8F393A400106000020E61000000400000001030000000100000005000000DB1EC3A02CAA65C0E08142F46F3132C0683A3658B1943740407608AE802056C038772CBE7A315840FE3F75DC546F5440D8703E4202F353C0A8DA191CAD625040DB1EC3A02CAA65C0E08142F46F3132C00103000000010000000A000000EA3F8EA591554DC068B0F14057A331C0BCBEB6BA13864040441E9C97655B56C064A1B4F607575B40D50BEC01A5CA51C05E3865B93DC2604062EB6F7C15F147C0C0BCDE771E553F407C5E4228F03733C02E993E5EE5F5634080431918DC6EFABF5679768A041454C094E6B75CCBFF4740BD1310CEFF3163C00C5CEC6C163F52408D2FC271938B65C0F8D6E868B7972940EA3F8EA591554DC068B0F14057A331C00103000000010000000C00000042EE1C4E713760C05463C74DDE4C53C0A9EBA313390653C0CADE4464936950C00057BE6AE64D1540E48E1FBAA9A455C040122B7073CA32406ACD1533ADBE50C000DB49CBC3BC1C401E71F5D75F9343C0F4AF7F62FFA05340BCF7EFB7C3C134C032988DAF9CDD634046A7AC44D72052400CC30E6DECB040C05808FE434E1548408ED43FC2EBC056C0D869DF363880534004E62B06695D49C018DAB1BD47D82D40B57B56F9A51465C0F03817F27340504042EE1C4E713760C05463C74DDE4C53C00103000000010000000B0000000522A260A9FB50C004B15A2C8E7B43C0E034AC71ED2D59C0ECC9E981274A50C0E8C3F900DEA130C0407AC72E308A2AC060871C3DB4F331C0C8EF5685F0F02EC06065F86AE91F2C4090FFA373EAB633C0A8448DE03B0C5140A1A880BE090044C022D45551459265409688A564DA665340F4C0C5D2070A4F40C4D131F73FEA534000FD0A0051CB2A4028A728827FF6354056E1076A182255C03E146B5092B555400522A260A9FB50C004B15A2C8E7B43C0':::GEOGRAPHY, (-0.7297516162474215):::FLOAT8, (-8332850917377672662):::INT8) :1]
panic: statement exec timeout


query
 SELECT st_buffer('01060000000100000001030000000100000007000000595466C414D8E9C7B8A9196ACC65E4C74A0576084BACDAC7E9A04CB1E011EFC77E701EEF7191E34748E7F0473F3AE8C782CF2146F24FE54706E38BB5CAA5D4C73A321CC4A1A4E047F86753B84FA4E647607EC493E1B7B7477A25F1A25773D1C7595466C414D8E9C7B8A9196ACC65E4C7':::GEOMETRY, 1.7976931348623157e+308:::FLOAT8, (-2791747071544515918):::INT8);
```

That seems like a malformed query possibly? We should either check if there's a bug in implementation, or else teach the Random Syntax tests to not create such queries.

--- a/pkg/geo/geomfn/buffer.go
+++ b/pkg/geo/geomfn/buffer.go
@@ -113,6 +113,13 @@ func ParseBufferParams(s string, distance float64) (BufferParams, float64, error
 
 // Buffer buffers a given Geometry by the supplied parameters.
 func Buffer(g geo.Geometry, params BufferParams, distance float64) (geo.Geometry, error) {
+	if params.p.QuadrantSegments < 1 {
+		return geo.Geometry{}, errors.Newf(
+			"must request at least 1 quadrant segment, requested %d quadrant segments",
+			params.p.QuadrantSegments,
+		)
+
+	}
 	if params.p.QuadrantSegments > geo.MaxAllowedSplitPoints {
 		return geo.Geometry{}, errors.Newf(
 			"attempting to split buffered geometry into too many quadrant segments; requested %d quadrant segments, max %d",

--- a/pkg/geo/geomfn/buffer_test.go
+++ b/pkg/geo/geomfn/buffer_test.go
@@ -102,4 +102,20 @@ func TestBuffer(t *testing.T) {
 			),
 		)
 	})
+
+	t.Run("too little quadrant segements", func(t *testing.T) {
+		for _, num := range []int{0, -1} {
+			t.Run(fmt.Sprintf("%d", num), func(t *testing.T) {
+				g := geo.MustParseGeometry("LINESTRING(0 0, 100 100)")
+				params, distance, err := ParseBufferParams(fmt.Sprintf("quad_segs=%d", num), 1000)
+				require.NoError(t, err)
+				_, err = Buffer(g, params, distance)
+				require.EqualError(
+					t,
+					err,
+					fmt.Sprintf("must request at least 1 quadrant segment, requested %d quadrant segments", num),
+				)
+			})
+		}
+	})
 }

